File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    int findFirst(vector<int>& nums, int t, int i, int j){
        int m = (i+j)/2;
        if(i>j) return -1;
        if(t>nums[m]){
            return findFirst(nums, t, m+1, j);
        }
        else if(t<nums[m]){
            return findFirst(nums, t, i, m-1);
        }
        else{
            while(m>=0 && nums[m]==t){
                m--;
        }
        }
        return m+1;
    }

    int findLast(vector<int>& nums, int t, int i, int j){
        int m = (i+j)/2;
        if(i>j) return -1;
        if(t>nums[m]){
            return findLast(nums, t, m+1, j);
        }
        else if(t<nums[m]){
            return findLast(nums, t, i, m-1);
        }
        else{
            while(m<nums.size() && nums[m]==t){
            m++;
        }
        }
        return m-1;
    }

    vector<int> searchRange(vector<int>& nums, int t) {
        vector<int> ans;
        int n = nums.size();
        ans.push_back(findFirst(nums, t, 0, n-1));
        ans.push_back(findLast(nums, t, 0, n-1));
        return ans;
    }
};

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int bSearchLeft(int[] nums, int left, int right, int target) {
        if(left > right) return -1;
        int mid = (left + right)/2;
        if(nums[mid] == target) {
            int index = bSearchLeft(nums, left, mid-1, target);
            return (index == -1)? mid: index;
        }
        if(target < nums[mid]) return bSearchLeft(nums, left, mid-1, target);
        return bSearchLeft(nums, mid+1, right, target);
    }
    public int bSearchRight(int[] nums, int left, int right, int target) {
        if(left > right) return -1;
        int mid = (left + right)/2;
        if(nums[mid] == target) {
            int index = bSearchRight(nums, mid+1, right, target);
            return (index == -1)? mid: index;
        }
        if(target < nums[mid]) return bSearchRight(nums, left, mid-1, target);
        return bSearchRight(nums, mid+1, right, target);
    }
    public int[] searchRange(int[] nums, int target) {
        int n = nums.length;
        int left = bSearchLeft(nums, 0, n-1, target);
        int right = bSearchRight(nums, 0, n-1, target);
        return new int[] {left, right};
    }
}